10up/animation-best-practices
# Feature: Sticky Header

See task https://tenup.teamwork.com/app/tasks/19284530



### Description of the Change


This PR adds a directionally aware sticky header demo.

This PR includes changes from PR #29

The script logic monitors the user's scroll position and conditionally hides a sticky header (moved off-canvas via `translateY()`) when the user has scrolled a configured minimum distance; default `24`.

A sticky header with a CSS state of `:hover`, `:focus`, or `:focus-within` will remain or become shown until that state is gone.

The script logic also...

- Sets and monitors `headerheight` on `resize` and redefines the CSS custom property `--sticky-header-height `
- Both which will affects the `scroll-margin-top` of a scroll `:target`

**Configurable Settings**
- JavaScript
    - `scrollPositionThreshold`
    How far the page needs to be scrolled for hide/show behavior to take affect (as a multiplier of the header height)
    - `scrollDistanceThreshold`
    How far a user needs to scroll to affect the hide/show behavior; default `24` (px)
- CSS
    - `--sticky-header-height`
    Default header height, primarily as a `.no-js` fallback; should match `.site-header` `height` with subsequent media 
queries values appropriate for the project
    - `--sticky-header-peek`
    The amount of sticky header _peeks_ out from the top of the viewport when the header is "hidden"; default `6px`
    - `--scroll-margin-top`
    The amount of scroll margin top applied to `:target` element; default `calc(1rem + var(--sticky-header-hide-peek))`
 

Includes #29 
Closes #24 

### How to test the Change


1. See demo, currently at [examples/ui/sticky-header/css](https://github.com/10up/animation-best-practices/tree/feature/sticky-header-19284530/src/examples/ui/sticky-header/css/)
2. See sticky header
3. See sticky header translates up (hidden) when the the user scrolls the minimum distance _down_
4. See sticky header translates down (shown) when the the user scrolls the minimum distance _up_
5. See sticky header is always shown while the user interacts with it (on `:hover` and `:hover` when scrolling)

### Changelog Entry


Added - New feature

### Credits
 @dylanjameswagner 

### Checklist:
